File: windows/net/samples/mux_mp4_avc_aac_file/Options.cs
/*
 *  Copyright (c) 2015 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/

using System;
using System.IO;
using System.Collections.Generic;
using CommandLine;
using CommandLine.Text;
using PrimoSoftware.AVBlocks;

/*
command line options:

	mux_mp4_avc_aac_file --input audio.mp4 video.mp4 --output output.mp4 [--fast-start]
*/

namespace MuxMp4AvcAacFileSample
{
    class Options
    {
        // mux_mp4_avc_aac_file options
        [Option('?', "help", HelpText = "Display this help screen")]
        public bool Help { get; set; }

        [OptionArray('i', "input", HelpText = "input file (aac, h264, mp4)")]
        public string[] InputFiles { get; set; }

        [Option('o', "output", HelpText = "output file")]
        public string OutputFile { get; set; }

        [Option('f', "fast-start", HelpText = "mp4 fast start")]
        public bool FastStart { get; set; }

        // The program parses the command line options and sets these properties
        public bool Error { get; private set; }

        string GetUsage()
        {
            return HelpText.AutoBuild(this, (HelpText current) => HelpText.DefaultParsingErrorsHandler(this, current));
        }

        void PrintUsage()
        {
            Console.WriteLine("Usage: mux_mp4_avc_aac_file --input audio.mp4 video.mp4 --output output.mp4 [--fast-start]");
            Console.WriteLine(GetUsage());
        }

        void ResetOptions()
        {
            Help = false;
            Error = false;
            InputFiles = null;
            OutputFile = null;
        }

        void SetDefaultOptions()
        {
            string exeDir = Path.GetDirectoryName(System.Reflection.Assembly.GetExecutingAssembly().Location);
            string inputFile = Path.Combine(exeDir, "..\\assets\\mov\\big_buck_bunny_trailer_iphone.m4v");

            InputFiles = new string[] { inputFile };
            OutputFile = Path.Combine(exeDir, "mux_mp4_avc_aac_file.mp4");

            Console.WriteLine("Using default options: --input " + inputFile + " --output " + OutputFile);
        }

        public bool Prepare(string[] args)
        {
            ResetOptions();

            if (args.Length == 0)
            {
                SetDefaultOptions();
            }
            else
            {
                if (!CommandLine.Parser.Default.ParseArguments(args, this))
                {
                    Console.WriteLine("Syntax error");
                    PrintUsage();
                    Error = true;
                    return false;
                }
            }

            if (Help)
            {
                PrintUsage();
                return false;
            }

            if (!Validate())
            {
                Console.WriteLine("\nRequired options are not set!");
                PrintUsage();
                Error = true;
                return false;
            }

            return true;
        }

        bool Validate()
        {
            bool res = true;

            Console.WriteLine("Input files: ");
            if ((InputFiles != null) && (InputFiles.Length > 0))
            {
                foreach (string s in InputFiles)
                {
                    Console.WriteLine("   " + s);
                }
            }
            else
            {
                Console.WriteLine("[not set]");
                res = false;
            }

            Console.Write("Output file: ");
            if (OutputFile == null)
            {
                Console.WriteLine("[not set]");
                res = false;
            }
            else
            {
                Console.WriteLine(OutputFile);
            }

            Console.WriteLine("Fast-start: " + (FastStart ? "yes" : "no"));

            return res;
        }
    }
}

File: macosx/samples/mux_mp4_remux_file/options.cpp
/*
 *  Copyright (c) 2016 Primo Software. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree.  
*/

#include <string>
#include <iostream>
#include <iomanip>
#include <sstream>
#include <assert.h>

#include <AVBlocks.h>
#include <PrimoReference++.h>

#include "options.h"
#include "program_options.h"
#include "util.h"

using namespace std;
using namespace primo::avblocks;
using namespace primo::codecs;

void help(primo::program_options::OptionsConfig<char>& optcfg)
{
    cout << "Usage: mux_mp4_remux_file --input <file.mp4> --output <output.mp4> [--fast-start]";
    cout << endl;
    primo::program_options::doHelp(cout, optcfg);
}

void setDefaultOptions(Options& opt)
{
    if (opt.output_file.empty())
    {
        opt.output_file = getExeDir() + "/mux_mp4_remux_file.mp4";
    }
    
    if (opt.input_file.empty())
    {
        opt.input_file = getExeDir() + "/../assets/mov/big_buck_bunny_trailer_iphone.m4v";
    }
}

bool validateOptions(Options& opt)
{
    if (opt.input_file.empty())
    {
        return false;
    }
    
    if (opt.output_file.empty())
    {
        return false;
    }
    return true;
}

ErrorCodes prepareOptions(Options &opt, int argc, char* argv[])
{
    if (argc < 2)
    {
        setDefaultOptions(opt);
        cout << "Using defaults:\n";
        cout << " --input " << opt.input_file;

        cout << " --output " << opt.output_file;
        cout << endl;
        return Parsed;
    }
    
    primo::program_options::OptionsConfig<char> optcfg;
    optcfg.addOptions()
    ("help,h",      opt.help,       "")
    ("input,i",     opt.input_file,  string(), "input file (mp4)")
    ("output,o",    opt.output_file, string(), "output file")
    ("fast-start,f",opt.fast_start, "mp4 fast start");
    
    try
    {
        primo::program_options::scanArgv(optcfg, argc, argv);
    }
    catch (primo::program_options::ParseFailure<char> &ex)
    {
        cout << ex.message() << endl;
        help(optcfg);
        return Error;
    }
    
    if (opt.help)
    {
        help(optcfg);
        return Command;
    }
    
    if (!validateOptions(opt))
    {
        help(optcfg);
        return Error;
    }
    
    return Parsed;
}
